File: Business/Business.php
<?php
/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 21:21
 */

interface BusinessService{
    public function doProcessing();
}

class EJBService  implements BusinessService{
    public function doProcessing()
    {
        // TODO: Implement doProcessing() method.
        var_dump("Processing task by invoking EJB Service");
    }
}


class JMSService  implements BusinessService{
    public function doProcessing()
    {
        // TODO: Implement doProcessing() method.
        var_dump("Processing task by invoking JMS Service");
    }
}


class BusinessLookUp{
    public function getBusinessService($serviceType){
        if($serviceType=="EJB"){
            return new EJBService();
        }else{
            return new JMSService();
        }

    }
}

class BusinessDelegate {
    private $lookupService;
    private $businessService;
    private $serviceType;
    public function __construct()
    {
        $this->lookupService=new BusinessLookUp();
    }

    public function setSevericeType($serviceType){
        $this->serviceType=$serviceType;
    }

    public function doTask(){
        $this->businessService=$this->lookupService->getBusinessService($this->serviceType);
        $this->businessService->doProcessing();
    }


}

class Client {
    private $businessService;
    public function __construct($BusinessDelegate)
    {
        $this->businessService=$BusinessDelegate;
    }
    public function doTask(){
        $this->businessService->doTask();
    }
}

//dmeo

$businessDelegate=new BusinessDelegate();
$client=new Client($businessDelegate);
$businessDelegate->setSevericeType("EJB");
$client->doTask();

$businessDelegate->setSevericeType("JMS");
$client->doTask();

File: Business/Business.go
package main

import "fmt"

type BusinessService interface {
	doProcessing()
}

type EJBService struct {

}

func(e *EJBService)doProcessing(){
	fmt.Println("Processing task by invoking EJB Service");
}

type JMSService struct {

}

func(e *JMSService)doProcessing(){
	fmt.Println("Processing task by invoking JMS Service");
}


type BusinessLookup struct {

}
func(b *BusinessLookup)getBusinessService(name string) BusinessService{
	if(name=="EJB"){
		return new(EJBService)
	}else{
		return new(JMSService)
	}
}


type BusinessDelegate struct {
	serviceType string
	businessLookup *BusinessLookup
}

func(b *BusinessDelegate)setSeviceType(serviceType string){
	b.serviceType=serviceType
}

func(b *BusinessDelegate)doTask(){
	service:=b.businessLookup.getBusinessService(b.serviceType)
	service.doProcessing()
}

type Client struct {
	businessDelegate *BusinessDelegate
}

func(c *Client)doTask(){
	c.businessDelegate.doTask()
}

func main(){
	businessDelegate:=&BusinessDelegate{"EJB",new(BusinessLookup)}
	client:=&Client{businessDelegate}
	client.doTask()

	businessDelegate.setSeviceType("JMS");
	client.doTask();

}

